### PR TITLE
[ELF] Add support for PT_OPENBSD_NOBTCFI

### PR DESCRIPTION
**Pull request by brad0**

### CODE REVIEW

File: lld/ELF/Driver.cpp [hunk 1]
```diff
@@ -1487,6 +1487,7 @@ static void readConfigs(Ctx &ctx, opt::InputArgList &args) {
       args, "keep-text-section-prefix", "nokeep-text-section-prefix", false);
   ctx.arg.zLrodataAfterBss =
       getZFlag(args, "lrodata-after-bss", "nolrodata-after-bss", false);
+  ctx.arg.zNoBtCfi = hasZOption(args, "nobtcfi");
```
Comment (MaskRay) [inline lld/ELF/Driver.cpp:1490]: Initialize the variable to false. Move this to setConfigs and only do it when osabi is openbsd.

**Review by MaskRay — COMMENTED** [review]
I think this is fine... but it is not a good idea to keep introducing new PT_OPENBSD program headers...

File: lld/ELF/Driver.cpp (hunk 1 above)
Comment (brad0) [reply lld/ELF/Driver.cpp]: I am not sure what you mean by initialize the variable to false.

Comment (brad0) [reply lld/ELF/Driver.cpp]: Also the comment at the top of setConfigs() doesn't really align with this being added there..
```
// Some Config members do not directly correspond to any particular
// command line options, but computed based on other Config values.
```

Comment (MaskRay) [reply lld/ELF/Driver.cpp]: This comment will be stale and can be updated.

I mean `bool zNoBtCfi = false;` in Config.h

File: lld/ELF/Driver.cpp [hunk 2]
```diff
@@ -1897,6 +1897,10 @@ static void setConfigs(Ctx &ctx, opt::InputArgList &args) {
       ErrAlways(ctx) << "cannot open --why-extract= file " << ctx.arg.whyExtract
                      << ": " << e.message();
   }
+
+  if (ctx.arg.osabi == ELFOSABI_OPENBSD) {
```
Comment (MaskRay) [inline lld/ELF/Driver.cpp:1901]: delete braces

File: lld/ELF/Writer.cpp [hunk 3]
```diff
@@ -2405,6 +2405,13 @@ Writer<ELFT>::createPhdrs(Partition &part) {
     addHdr(PT_GNU_STACK, perm)->p_memsz = ctx.arg.zStackSize;
   }
 
+  if (ctx.arg.osabi == ELFOSABI_OPENBSD) {
```
Comment (MaskRay) [inline lld/ELF/Writer.cpp:2408]: The osabi check can be deleted as zNoBtCfi cannot be true for other osabi values

File: lld/test/ELF/openbsd-phdr.s [hunk 4]
```diff
@@ -4,6 +4,10 @@
 # RUN: ld.lld randomdata.o -o randomdata
 # RUN: llvm-readelf -S -l randomdata | FileCheck %s --check-prefix=RANDOMDATA
 
+# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-openbsd /dev/null -o nobtcfi.o
+# RUN: ld.lld -z nobtcfi nobtcfi.o -o nobtcfi
+# RUN: llvm-readelf -l nobtcfi | FileCheck %s --check-prefix=NOBTCFI
+
 # RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-openbsd /dev/null -o wxneeded.o
```
Comment (MaskRay) [inline lld/test/ELF/openbsd-phdr.s:11]: we can rename `wxneeded.o` to `empty.o` and reuse it for `empty.o -o nobtcfi`

**Review by MaskRay — APPROVED** [review]

File: lld/test/ELF/openbsd-phdr.s (hunk 4 above)
Comment (brad0) [reply lld/test/ELF/openbsd-phdr.s:11]:
```
# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-openbsd /dev/null -o empty.o
# RUN: ld.lld -z nobtcfi empty.o -o nobtcfi
# RUN: llvm-readelf -l nobtcfi | FileCheck %s --check-prefix=NOBTCFI

# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-openbsd /dev/null -o empty.o
# RUN: ld.lld -z wxneeded empty.o -o wxneeded
# RUN: llvm-readelf -l wxneeded | FileCheck %s --check-prefix=WXNEEDED
```

You mean like that?

Comment (MaskRay) [reply lld/test/ELF/openbsd-phdr.s:11]: Delete the second `# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-openbsd /dev/null -o empty.o`

File: lld/ELF/Driver.cpp [hunk 5]
```diff
@@ -1897,6 +1897,9 @@ static void setConfigs(Ctx &ctx, opt::InputArgList &args) {
       ErrAlways(ctx) << "cannot open --why-extract= file " << ctx.arg.whyExtract
                      << ": " << e.message();
   }
+
+  if (ctx.arg.osabi == ELFOSABI_OPENBSD)
```
Comment (MaskRay) [inline lld/ELF/Driver.cpp:1901]: `ld.lld -z nobtcfi empty.o` has a false positive warning. Perhaps we need to move this variable back to readConfigs.

**Pull request merged**